File: android/app/src/main/kotlin/com/dexwritescode/workout/ui/workout/ActiveWorkoutViewModel.kt
package com.dexwritescode.workout.ui.workout

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.dexwritescode.workout.data.db.dao.CompletedExerciseDao
import com.dexwritescode.workout.data.db.dao.ExerciseDao
import com.dexwritescode.workout.data.db.dao.ExerciseSetDao
import com.dexwritescode.workout.data.db.dao.MuscleRecoveryStateDao
import com.dexwritescode.workout.data.db.dao.TemplateExerciseDao
import com.dexwritescode.workout.data.db.dao.TemplateSetDao
import com.dexwritescode.workout.data.db.dao.WorkoutSessionDao
import com.dexwritescode.workout.data.db.dao.WorkoutTemplateDao
import com.dexwritescode.workout.data.model.entity.CompletedExercise
import com.dexwritescode.workout.data.model.entity.Exercise
import com.dexwritescode.workout.data.model.entity.ExerciseSet
import com.dexwritescode.workout.data.model.entity.TemplateExercise
import com.dexwritescode.workout.data.model.entity.TemplateSet
import com.dexwritescode.workout.data.model.entity.WorkoutSession
import com.dexwritescode.workout.data.model.entity.WorkoutTemplate
import com.dexwritescode.workout.data.model.enums.MuscleGroup
import com.dexwritescode.workout.data.services.model.CompletedExerciseDetail
import com.dexwritescode.workout.data.services.model.WorkoutSessionDetail
import com.dexwritescode.workout.data.services.recovery.RecoveryEngine
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

// One slot in the active workout — pairs template metadata with live session data.
data class ExerciseSlot(
    val templateExercise: TemplateExercise,
    val exercise: Exercise?,
    val templateSets: List<TemplateSet>,
    val completedExercise: CompletedExercise,
    val completedSets: List<ExerciseSet>,
    val isAdhoc: Boolean = false,
)

data class WorkoutSummary(
    val durationMs: Long,
    val exercisesCompleted: Int,
    val totalExercises: Int,
    val totalSets: Int,
    val totalVolume: Double,
    val exerciseBreakdowns: List<ExerciseBreakdown>,
    val musclesWorked: Map<MuscleGroup, Double>,
) {
    data class ExerciseBreakdown(val name: String, val sets: List<SetDetail>)
    data class SetDetail(val setNumber: Int, val weight: Double, val reps: Int)
}

sealed class WorkoutState {
    object Loading : WorkoutState()
    data class NotStarted(val template: WorkoutTemplate, val slots: List<ExerciseSlot>) : WorkoutState()
    data class InProgress(
        val template: WorkoutTemplate,
        val session: WorkoutSession,
        val slots: List<ExerciseSlot>,
        val currentIndex: Int,
        val elapsedSeconds: Long,
    ) : WorkoutState()
    data class Finished(val session: WorkoutSession, val summary: WorkoutSummary) : WorkoutState()
}

class ActiveWorkoutViewModel(
    private val templateId: String,
    private val templateDao: WorkoutTemplateDao,
    private val templateExerciseDao: TemplateExerciseDao,
    private val templateSetDao: TemplateSetDao,
    private val exerciseDao: ExerciseDao,
    private val sessionDao: WorkoutSessionDao,
    private val completedExerciseDao: CompletedExerciseDao,
    private val exerciseSetDao: ExerciseSetDao,
    private val recoveryDao: MuscleRecoveryStateDao,
) : ViewModel() {

    private val _state = MutableStateFlow<WorkoutState>(WorkoutState.Loading)
    val state: StateFlow<WorkoutState> = _state.asStateFlow()

    private var sessionStartTime: Long = 0L

    init {
        viewModelScope.launch { loadTemplate() }
    }

    private suspend fun loadTemplate() {
        val template = templateDao.getById(templateId) ?: return
        val templateExercises = templateExerciseDao.getByTemplate(templateId).first()
        val slots = buildSlotsFromTemplateExercises(templateExercises.sortedBy { it.order })
        _state.value = WorkoutState.NotStarted(template, slots)
    }

    private suspend fun buildSlotsFromTemplateExercises(
        templateExercises: List<TemplateExercise>,
        adhocFlags: Set<String> = emptySet(),
    ): List<ExerciseSlot> = templateExercises.map { te ->
        val exercise = te.exerciseId?.let { exerciseDao.getById(it) }
        val templateSets = templateSetDao.getByTemplateExercise(te.id).first().sortedBy { it.order }
        ExerciseSlot(
            templateExercise = te,
            exercise = exercise,
            templateSets = templateSets,
            completedExercise = CompletedExercise(sessionId = null, exerciseId = te.exerciseId, order = te.order),
            completedSets = emptyList(),
            isAdhoc = te.id in adhocFlags,
        )
    }

    // MARK: - Session Lifecycle

    fun startWorkout() {
        val current = _state.value as? WorkoutState.NotStarted ?: return
        viewModelScope.launch {
            val session = WorkoutSession(templateId = templateId)
            sessionDao.insert(session)
            sessionStartTime = session.startTime
            templateDao.update(current.template.copy(lastUsedDate = System.currentTimeMillis()))

            val updatedSlots = current.slots.map { slot ->
                val ce = CompletedExercise(
                    sessionId = session.id,
                    exerciseId = slot.templateExercise.exerciseId,
                    order = slot.templateExercise.order,
                )
                completedExerciseDao.insert(ce)
                slot.copy(completedExercise = ce)
            }

            _state.value = WorkoutState.InProgress(
                template = current.template,
                session = session,
                slots = updatedSlots,
                currentIndex = 0,
                elapsedSeconds = 0L,
            )
            startElapsedTimer()
        }
    }

    fun finishWorkout() {
        val current = _state.value as? WorkoutState.InProgress ?: return
        viewModelScope.launch {
            val endTime = System.currentTimeMillis()
            val finished = current.session.copy(endTime = endTime, isCompleted = true)
            sessionDao.update(finished)
            val summary = buildSummary(finished, current.slots)
            _state.value = WorkoutState.Finished(session = finished, summary = summary)
        }
    }

    fun cancelWorkout() {
        val current = _state.value as? WorkoutState.InProgress ?: return
        viewModelScope.launch {
            current.slots.filter { it.isAdhoc }.forEach { templateExerciseDao.delete(it.templateExercise) }
            sessionDao.delete(current.session)
            val slots = buildSlotsFromTemplateExercises(
                current.slots.filter { !it.isAdhoc }.map { it.templateExercise }.sortedBy { it.order }
            )
            _state.value = WorkoutState.NotStarted(current.template, slots)
        }
    }

    fun addExercise(exercise: Exercise) {
        val current = _state.value as? WorkoutState.InProgress ?: return
        viewModelScope.launch {
            val order = current.slots.size
            val te = TemplateExercise(
                templateId = null,
                exerciseId = exercise.id,
                order = order,
                targetSets = 3,
                targetReps = 10,
                restSeconds = 90,
            )
            templateExerciseDao.insert(te)

            val ce = CompletedExercise(sessionId = current.session.id, exerciseId = exercise.id, order = order)
            completedExerciseDao.insert(ce)

            val newSlot = ExerciseSlot(
                templateExercise = te,
                exercise = exercise,
                templateSets = emptyList(),
                completedExercise = ce,
                completedSets = emptyList(),
                isAdhoc = true,
            )
            _state.value = current.copy(slots = current.slots + newSlot)
        }
    }

    fun removeExercise(index: Int) {
        val current = _state.value as? WorkoutState.InProgress ?: return
        if (index !in current.slots.indices) return
        viewModelScope.launch {
            val slot = current.slots[index]
            completedExerciseDao.delete(slot.completedExercise)
            if (slot.isAdhoc) templateExerciseDao.delete(slot.templateExercise)
            val updatedSlots = current.slots.toMutableList().also { it.removeAt(index) }
            val newIndex = minOf(current.currentIndex, maxOf(0, updatedSlots.size - 1))
            _state.value = current.copy(slots = updatedSlots, currentIndex = newIndex)
        }
    }

    fun selectExercise(index: Int) {
        val current = _state.value as? WorkoutState.InProgress ?: return
        if (index !in current.slots.indices) return
        _state.value = current.copy(currentIndex = index)
    }

    fun markExerciseComplete(index: Int) {
        val current = _state.value as? WorkoutState.InProgress ?: return
        val next = current.slots.indices.drop(index + 1).firstOrNull { i ->
            val slot = current.slots[i]
            slot.completedSets.count { it.isCompleted } < slot.templateExercise.targetSets
        }
        if (next != null) _state.value = current.copy(currentIndex = next)
    }

    // MARK: - Set Logging

    fun logSet(slotIndex: Int, weight: Double, reps: Int) {
        val current = _state.value as? WorkoutState.InProgress ?: return
        if (slotIndex !in current.slots.indices) return
        viewModelScope.launch {
            val slot = current.slots[slotIndex]
            val setNumber = slot.completedSets.count { it.isCompleted } + 1
            val newSet = ExerciseSet(
                completedExerciseId = slot.completedExercise.id,
                setNumber = setNumber,
                weight = weight,
                reps = reps,
                isCompleted = true,
                completedAt = System.currentTimeMillis(),
            )
            exerciseSetDao.insert(newSet)
            val updatedSlots = current.slots.toMutableList()
            updatedSlots[slotIndex] = slot.copy(completedSets = slot.completedSets + newSet)
            _state.value = current.copy(slots = updatedSlots)
        }
    }

    fun updateSet(slotIndex: Int, set: ExerciseSet, weight: Double, reps: Int) {
        val current = _state.value as? WorkoutState.InProgress ?: return
        if (slotIndex !in current.slots.indices) return
        viewModelScope.launch {
            val updated = set.copy(weight = weight, reps = reps)
            exerciseSetDao.update(updated)
            val updatedSlots = current.slots.toMutableList()
            val slot = updatedSlots[slotIndex]
            updatedSlots[slotIndex] = slot.copy(
                completedSets = slot.completedSets.map { if (it.id == set.id) updated else it }
            )
            _state.value = current.copy(slots = updatedSlots)
        }
    }

    fun deleteSet(slotIndex: Int, set: ExerciseSet) {
        val current = _state.value as? WorkoutState.InProgress ?: return
        if (slotIndex !in current.slots.indices) return
        viewModelScope.launch {
            exerciseSetDao.delete(set)
            val updatedSlots = current.slots.toMutableList()
            val slot = updatedSlots[slotIndex]
            val remaining = slot.completedSets
                .filter { it.id != set.id }
                .mapIndexed { i, s -> s.copy(setNumber = i + 1) }
            remaining.forEach { exerciseSetDao.update(it) }
            updatedSlots[slotIndex] = slot.copy(completedSets = remaining)
            _state.value = current.copy(slots = updatedSlots)
        }
    }

    // MARK: - Save / Discard

    fun saveWorkout(notes: String?) {
        val current = _state.value as? WorkoutState.Finished ?: return
        viewModelScope.launch {
            val updated = current.session.copy(notes = notes?.takeIf { it.isNotBlank() })
            sessionDao.update(updated)
            val allCompleted = completedExerciseDao.getBySessionOnce(updated.id)
            val details = allCompleted.mapNotNull { ce ->
                ce.exerciseId?.let { exerciseDao.getById(it) }?.let { ex ->
                    CompletedExerciseDetail(ce, ex, exerciseSetDao.getByCompletedExerciseOnce(ce.id))
                }
            }
            val detail = WorkoutSessionDetail(session = updated, completedExercises = details)
            RecoveryEngine.updateRecoveryStates(detail, recoveryDao)
        }
    }

    fun discardWorkout() {
        val current = _state.value as? WorkoutState.Finished ?: return
        viewModelScope.launch { sessionDao.delete(current.session) }
    }

    // MARK: - Helpers

    private fun startElapsedTimer() {
        viewModelScope.launch {
            while (true) {
                delay(1_000)
                val current = _state.value as? WorkoutState.InProgress ?: break
                _state.value = current.copy(
                    elapsedSeconds = (System.currentTimeMillis() - sessionStartTime) / 1000
                )
            }
        }
    }

    private suspend fun buildSummary(session: WorkoutSession, slots: List<ExerciseSlot>): WorkoutSummary {
        val allCompleted = completedExerciseDao.getBySessionOnce(session.id)
        val allSets = allCompleted.flatMap { exerciseSetDao.getByCompletedExerciseOnce(it.id) }
            .filter { it.isCompleted }

        val details = allCompleted.mapNotNull { ce ->
            ce.exerciseId?.let { exerciseDao.getById(it) }?.let { ex ->
                CompletedExerciseDetail(ce, ex, exerciseSetDao.getByCompletedExerciseOnce(ce.id))
            }
        }
        val musclesWorked = RecoveryEngine.calculateFatigueDeltas(
            WorkoutSessionDetail(session = session, completedExercises = details)
        )

        val breakdowns = slots.mapNotNull { slot ->
            val done = slot.completedSets.filter { it.isCompleted }.sortedBy { it.setNumber }
            if (done.isEmpty()) return@mapNotNull null
            WorkoutSummary.ExerciseBreakdown(
                name = slot.exercise?.name ?: "Unknown",
                sets = done.map { WorkoutSummary.SetDetail(it.setNumber, it.weight, it.reps) },
            )
        }

        return WorkoutSummary(
            durationMs = session.durationMs ?: 0L,
            exercisesCompleted = slots.count { it.completedSets.any { s -> s.isCompleted } },
            totalExercises = slots.size,
            totalSets = allSets.size,
            totalVolume = allSets.sumOf { it.weight * it.reps },
            exerciseBreakdowns = breakdowns,
            musclesWorked = musclesWorked,
        )
    }

    companion object {
        fun formatDuration(seconds: Long): String {
            val h = seconds / 3600
            val m = (seconds % 3600) / 60
            val s = seconds % 60
            return if (h > 0) "%d:%02d:%02d".format(h, m, s) else "%d:%02d".format(m, s)
        }
    }

    class Factory(
        private val templateId: String,
        private val templateDao: WorkoutTemplateDao,
        private val templateExerciseDao: TemplateExerciseDao,
        private val templateSetDao: TemplateSetDao,
        private val exerciseDao: ExerciseDao,
        private val sessionDao: WorkoutSessionDao,
        private val completedExerciseDao: CompletedExerciseDao,
        private val exerciseSetDao: ExerciseSetDao,
        private val recoveryDao: MuscleRecoveryStateDao,
    ) : ViewModelProvider.Factory {
        @Suppress("UNCHECKED_CAST")
        override fun <T : ViewModel> create(modelClass: Class<T>): T = ActiveWorkoutViewModel(
            templateId, templateDao, templateExerciseDao, templateSetDao,
            exerciseDao, sessionDao, completedExerciseDao, exerciseSetDao, recoveryDao,
        ) as T
    }
}

File: Workout/Views/ActiveWorkout/ActiveWorkoutViewModel.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
//
//  ActiveWorkoutViewModel.swift
//  Workout
//
//  Created by Dexter Darwich on 2025-12-30.
//

import Foundation
import SwiftData

/// Manages the lifecycle of an active workout session
@Observable
final class ActiveWorkoutViewModel {
    
    // MARK: - State
    
    enum WorkoutState: Equatable {
        case notStarted
        case inProgress
        case finished
    }
    
    private(set) var state: WorkoutState = .notStarted
    private(set) var session: WorkoutSession?
    private(set) var currentExerciseIndex: Int = 0
    private(set) var sessionExercises: [TemplateExercise] = []
    private(set) var adhocExercises: [TemplateExercise] = []

    let template: WorkoutTemplate?
    let sessionName: String
    private let modelContext: ModelContext

    /// Sorted exercises for the active session — always the session's own copy (see
    /// `startWorkout()`), never a live read of the template/draft's current exercises.
    var sortedExercises: [TemplateExercise] {
        sessionExercises
    }

    /// All exercises for the active session — session-local copy of template exercises
    /// plus any exercises added mid-workout. Does not mutate the template.
    var allTemplateExercises: [TemplateExercise] {
        sessionExercises + adhocExercises
    }

    // MARK: - Init

    init(template: WorkoutTemplate, modelContext: ModelContext) {
        self.template = template
        self.sessionName = template.name
        self.modelContext = modelContext
    }

    // MARK: - Session Lifecycle
    
    /// Starts a new workout session, creating CompletedExercise entries for each template exercise
    func startWorkout() {
        let isDraftSession = template?.draftKind == .workoutStaging
        let newSession = WorkoutSession(template: isDraftSession ? nil : template)
        if isDraftSession || template == nil { newSession.sessionTitle = sessionName }
        modelContext.insert(newSession)

        // Deep-copy the template/draft's exercises into new, unattached rows (te.template = nil,
        // same convention addExercise() uses below) rather than holding live references. This
        // decouples the running session from the draft entirely — editing/regenerating/loading
        // the draft while this workout is active can't corrupt what's being tracked here.
        if let template {
            sessionExercises = template.exercises
                .sorted { $0.order < $1.order }
                .map { $0.duplicated(in: modelContext) }
        }

        for templateExercise in sessionExercises {
            let completed = CompletedExercise(order: templateExercise.order)
            completed.exercise = templateExercise.exercise
            completed.session = newSession
            modelContext.insert(completed)
            newSession.completedExercises.append(completed)
        }

        if let sourceID = template?.sourceTemplateID,
           let source = WorkoutTemplate.template(withID: sourceID, in: modelContext) {
            source.lastUsedDate = Date()
            source.timesStarted += 1
        }

        session = newSession
        currentExerciseIndex = 0
        state = .inProgress
    }
    
    /// Finishes the current workout session
    func finishWorkout() {
        guard let session, state == .inProgress else { return }
        session.endTime = Date()
        session.isCompleted = true
        state = .finished
    }
    
    /// Adds an exercise to the current session without modifying the template
    func addExercise(_ exercise: Exercise) {
        guard let session else { return }
        let order = allTemplateExercises.count
        let te = TemplateExercise(order: order, targetSets: 3, targetReps: 10, restSeconds: 0)
        te.exercise = exercise
        modelContext.insert(te)
        adhocExercises.append(te)

        let ce = CompletedExercise(order: order)
        ce.exercise = exercise
        ce.session = session
        modelContext.insert(ce)
        session.completedExercises.append(ce)
    }

    /// Removes an exercise from the current session by its position in allTemplateExercises.
    /// Template exercises are removed from the session list only — the template is not modified.
    func removeExercise(at index: Int) {
        guard let session else { return }
        let sorted = sortedCompletedExercises
        guard index < sorted.count else { return }

        let ce = sorted[index]
        session.completedExercises.removeAll { $0 === ce }
        modelContext.delete(ce)

        if index < sessionExercises.count {
            sessionExercises.remove(at: index)
        } else {
            let adhocIndex = index - sessionExercises.count
            if adhocIndex < adhocExercises.count {
                modelContext.delete(adhocExercises.remove(at: adhocIndex))
            }
        }

        for (i, c) in sortedCompletedExercises.enumerated() { c.order = i }

        let remaining = allTemplateExercises.count
        if currentExerciseIndex >= remaining { currentExerciseIndex = max(0, remaining - 1) }
    }

    /// Cancels and deletes the current workout session
    func cancelWorkout() {
        // sessionExercises/adhocExercises are the session's own independent copies (see
        // startWorkout()) — safe to delete regardless of outcome. The draft's own exercises are
        // left untouched: nothing was completed, so the staged plan should still be there to retry.
        for te in sessionExercises { modelContext.delete(te) }
        for te in adhocExercises { modelContext.delete(te) }
        adhocExercises = []
        sessionExercises = []
        guard let session else { return }
        modelContext.delete(session)
        self.session = nil
        state = .notStarted
    }
    
    // MARK: - Navigation
    
    /// Moves to the next exercise in the list
    func moveToNextExercise() {
        let maxIndex = sortedExercises.count - 1
        if currentExerciseIndex < maxIndex {
            currentExerciseIndex += 1
        }
    }
    
    /// Moves to the previous exercise in the list
    func moveToPreviousExercise() {
        if currentExerciseIndex > 0 {
            currentExerciseIndex -= 1
        }
    }
    
    /// Selects a specific exercise by index
    func selectExercise(at index: Int) {
        guard index >= 0 && index < sortedExercises.count else { return }
        currentExerciseIndex = index
    }
    
    // MARK: - Set Completion
    
    /// Marks an exercise as complete and auto-advances to the next incomplete exercise
    func markExerciseComplete(at index: Int) {
        let all = allTemplateExercises
        let maxIndex = all.count - 1
        guard index + 1 <= maxIndex else { return }
        for i in (index + 1)...maxIndex {
            let te = all[i]
            let ce = sortedCompletedExercises.count > i ? sortedCompletedExercises[i] : nil
            let done = ce?.sets.filter(\.isCompleted).count ?? 0
            if done < te.targetSets {
                currentExerciseIndex = i
                return
            }
        }
        // If all after are done, stay put
    }
    
    // MARK: - Summary Stats
    
    struct WorkoutSummary {
        let duration: TimeInterval
        let exercisesCompleted: Int
        let totalExercises: Int
        let totalSets: Int
        let totalVolume: Double // weight × reps summed
        let exerciseBreakdowns: [ExerciseBreakdown]
        let musclesWorked: [MuscleGroup: Double] // muscle → fatigue delta
    }
    
    struct ExerciseBreakdown {
        let name: String
        let sets: [SetDetail]
    }
    
    struct SetDetail {
        let setNumber: Int
        let weight: Double
        let reps: Int
    }
    
    /// Computes summary stats for the finished workout
    var summary: WorkoutSummary? {
        guard let session, state == .finished else { return nil }
        
        let duration = session.duration ?? 0
        let completedExercises = sortedCompletedExercises
        
        let exercisesWithSets = completedExercises.filter { !$0.sets.filter(\.isCompleted).isEmpty }
        let allCompletedSets = completedExercises.flatMap { $0.sets.filter(\.isCompleted) }
        let totalVolume = allCompletedSets.reduce(0.0) { $0 + $1.weight * Double($1.reps) }
        
        let breakdowns: [ExerciseBreakdown] = completedExercises.compactMap { ce in
            let doneSets = ce.sets.filter(\.isCompleted).sorted { $0.setNumber < $1.setNumber }
            guard !doneSets.isEmpty else { return nil }
            return ExerciseBreakdown(
                name: ce.exercise?.name ?? "Unknown",
                sets: doneSets.map { SetDetail(setNumber: $0.setNumber, weight: $0.weight, reps: $0.reps) }
            )
        }
        
        let musclesWorked = RecoveryEngine.calculateFatigueDeltas(for: session)
        
        return WorkoutSummary(
            duration: duration,
            exercisesCompleted: exercisesWithSets.count,
            totalExercises: completedExercises.count,
            totalSets: allCompletedSets.count,
            totalVolume: totalVolume,
            exerciseBreakdowns: breakdowns,
            musclesWorked: musclesWorked
        )
    }
    
    /// Saves the workout and updates muscle recovery states
    func saveWorkout() {
        guard let session else { return }
        session.isCompleted = true
        RecoveryEngine.updateRecoveryStates(for: session, modelContext: modelContext)

        // sessionExercises/adhocExercises are independent copies with no further purpose now
        // that CompletedExercise/ExerciseSet hold the real, saved record.
        for te in sessionExercises { modelContext.delete(te) }
        for te in adhocExercises { modelContext.delete(te) }
        sessionExercises = []
        adhocExercises = []

        // The draft's own exercises are cleared only on an actual completion, so the next time
        // the staging view appears it starts from a clean slate rather than an already-used plan.
        if let template, template.draftKind == .workoutStaging {
            template.clearExercises(context: modelContext)
            template.sourceTemplateID = nil
            template.autoLoadedForWeekday = nil
        }
    }

    /// Discards the workout by deleting the session
    func discardWorkout() {
        // Same reasoning as cancelWorkout(): these are independent copies, safe to delete
        // regardless of outcome, but the draft's own exercises are left untouched since nothing
        // was actually completed.
        for te in sessionExercises { modelContext.delete(te) }
        for te in adhocExercises { modelContext.delete(te) }
        adhocExercises = []
        sessionExercises = []
        guard let session else { return }
        modelContext.delete(session)
        self.session = nil
    }
    
    /// Updates session notes
    func updateNotes(_ notes: String) {
        session?.notes = notes.isEmpty ? nil : notes
    }
    
    // MARK: - Helpers
    
    /// Returns the sorted completed exercises matching the template order
    var sortedCompletedExercises: [CompletedExercise] {
        session?.completedExercises.sorted { $0.order < $1.order } ?? []
    }
    
    /// Elapsed time since workout started (for use with TimelineView)
    var elapsedTime: TimeInterval {
        guard let session, state == .inProgress else { return 0 }
        return Date().timeIntervalSince(session.startTime)
    }
    
    /// Formats a time interval as MM:SS or H:MM:SS
    static func formatDuration(_ interval: TimeInterval) -> String {
        let totalSeconds = Int(interval)
        let hours = totalSeconds / 3600
        let minutes = (totalSeconds % 3600) / 60
        let seconds = totalSeconds % 60
        
        if hours > 0 {
            return String(format: "%d:%02d:%02d", hours, minutes, seconds)
        } else {
            return String(format: "%d:%02d", minutes, seconds)
        }
    }
}
